provisioner/terraform/provision.go

```diff
@@ -298,141 +136,60 @@ func (t *terraform) Provision(stream proto.DRPCProvisioner_ProvisionStream) erro
-// path and reads the merged state. If there is no state, `os.ErrNotExist` will
-// be returned.
-func pullTerraformState(ctx context.Context, terraform *tfexec.Terraform, statefilePath string) (*tfjson.State, error) {
-	statefile, err := os.OpenFile(statefilePath, os.O_CREATE|os.O_RDWR, 0600)
-	if err != nil {
-		return nil, xerrors.Errorf("open statefile %q: %w", statefilePath, err)
-	}
-	defer statefile.Close()
-
-	// #nosec
-	cmd := exec.CommandContext(ctx, terraform.ExecPath(), "state", "pull")
-	cmd.Dir = terraform.WorkingDir()
-	cmd.Stdout = statefile
-	err = cmd.Run()
-	if err != nil {
-		return nil, xerrors.Errorf("pull terraform state: %w", err)
+func provisionEnv(start *proto.Provision_Start) ([]string, error) {
+	env := os.Environ()
+	env = append(env,
+		"CODER_AGENT_URL="+start.Metadata.CoderUrl,
+		"CODER_WORKSPACE_TRANSITION="+strings.ToLower(start.Metadata.WorkspaceTransition.String()),
+		"CODER_WORKSPACE_NAME="+start.Metadata.WorkspaceName,
+		"CODER_WORKSPACE_OWNER="+start.Metadata.WorkspaceOwner,
+		"CODER_WORKSPACE_ID="+start.Metadata.WorkspaceId,
+		"CODER_WORKSPACE_OWNER_ID="+start.Metadata.WorkspaceOwnerId,
+	)
+	for key, value := range provisionersdk.AgentScriptEnv() {
+		env = append(env, key+"="+value)
 	}
-
-	state, err := terraform.ShowStateFile(ctx, statefilePath)
-	if err != nil {
-		if noStateRegex.MatchString(err.Error()) {
-			return nil, os.ErrNotExist
+	for _, param := range start.ParameterValues {
+		switch param.DestinationScheme {
+		case proto.ParameterDestination_ENVIRONMENT_VARIABLE:
+			env = append(env, fmt.Sprintf("%s=%s", param.Name, param.Value))
+		case proto.ParameterDestination_PROVISIONER_VARIABLE:
+			continue
+		default:
+			return nil, xerrors.Errorf("unsupported parameter type %q for %q", param.DestinationScheme, param.Name)
 		}
-
-		return nil, xerrors.Errorf("show terraform state: %w", err)
 	}
-
-	return state, nil
-}
-
-type terraformProvisionLog struct {
-	Level   string `json:"@level"`
-	Message string `json:"@message"`
-
-	Diagnostic *terraformProvisionLogDiagnostic `json:"diagnostic"`
-}
-
-type terraformProvisionLogDiagnostic struct {
-	Severity string `json:"severity"`
-	Summary  string `json:"summary"`
-	Detail   string `json:"detail"`
+	return env, nil
 }
 
-func convertTerraformLogLevel(logLevel string) (proto.LogLevel, error) {
-	switch strings.ToLower(logLevel) {
-	case "trace":
-		return proto.LogLevel_TRACE, nil
-	case "debug":
-		return proto.LogLevel_DEBUG, nil
-	case "info":
-		return proto.LogLevel_INFO, nil
-	case "warn":
-		return proto.LogLevel_WARN, nil
-	case "error":
-		return proto.LogLevel_ERROR, nil
-	default:
-		return proto.LogLevel(0), xerrors.Errorf("invalid log level %q", logLevel)
+func logTerraformEnvVars(logger provisionersdk.Logger) error {
+	env := os.Environ()
+	for _, e := range env {
+		if strings.HasPrefix(e, "TF_") {
+			err := logger.Log(&proto.Log{Level: proto.LogLevel_WARN, Output: "terraform environment variable: " + e})
```

Comment:
Terraform variables can be applied with [environment variables](https://www.terraform.io/cli/config/environment-variables#tf_var_name), which means logs would leak secrets. We could `SplitN` by `=` to print the keys instead.